In [1]:
# Check for changes in .py files every time a cell is ran
%load_ext autoreload
%autoreload 2

In [2]:
import sys
import os

# This adds the parent directory (project root) to the search path
sys.path.append(os.path.abspath(".."))

In [3]:
# Import function from database.py for sql files executing
from src.database import execute_sql_file
from pathlib import Path

In [4]:
# Testing the function
# Resolve the path to the SQL file relative to the notebook
query_path = Path("../queries/00_get_tables.sql")

# Execute query and store results
tables = execute_sql_file(query_path)

# Print results
print(tables)

[('countries',), ('grapes',), ('wineries',), ('flavor_groups',), ('keywords',), ('regions',), ('most_used_grapes_per_country',), ('toplists',), ('wines',), ('vintages',), ('keywords_wine',), ('vintage_toplists_rankings',)]


## Business questions

### 1. We want to highlight 10 wines to increase our sales. Which ones should we choose and why?

In [5]:
query_path = Path("../queries/01_highlight_10_wines.sql") 

top_10_wines = execute_sql_file(query_path)

for wine, ratings_avg, ratings_count in top_10_wines:
    print(f"{wine=}, {ratings_avg=}, {ratings_count=}")


wine='Unico', ratings_avg=4.7, ratings_count=45140
wine='Sauternes', ratings_avg=4.7, ratings_count=44126
wine='Special Selection Cabernet Sauvignon', ratings_avg=4.7, ratings_count=41236
wine='Pomerol', ratings_avg=4.7, ratings_count=32157
wine='Grange', ratings_avg=4.7, ratings_count=24356
wine='Toscana', ratings_avg=4.7, ratings_count=16284
wine='Unico Reserva Especial Edición', ratings_avg=4.7, ratings_count=13025
wine='Almaviva', ratings_avg=4.6, ratings_count=48075
wine='Bolgheri Superiore', ratings_avg=4.6, ratings_count=47939
wine='Château Margaux (Premier Grand Cru Classé)', ratings_avg=4.6, ratings_count=45350


Visualization process

Created a visualizations.py script containing all visualization functions


1.a 1st question visualization

Scatter Plot:

X-Axis: Number of Ratings (ratings_count)

Y-Axis: Average Rating (ratings_average)

Background: The rest of the database plotted as faint, transparent markers to show the overarching trend.

The Highlight: Our top 10 wines plotted as prominent, colored markers inside our defined "Sweet Spot" box.

In [6]:
# Import and call 1st question visual function
from src.visualizations import plot_top_10_bar

fig1 = plot_top_10_bar(top_10_wines)
fig1.show()

### Here is the exact framework for a data story:

1. The Goal (The Hook): Acknowledge what the business wants (e.g., increase sales).
2. The Strategy (The Pivot): Explain how you manipulated the data without using technical terms.
3. The Result (The Payoff): Present the solution and explicitly state the business value.

1.b Explanation for stakeholders

To drive new sales, we must avoid spending our marketing budget on over-saturated 'mega-brands' that our customers already know and buy. Instead, we targeted a 'Sweet Spot' of 10 Hidden Gems. These are wines that have definitively proven their elite quality across thousands of customer reviews, but have not yet peaked in mainstream fame. By highlighting these specific wines, we guarantee we are introducing highly reliable, top-tier products to a fresh audience, which maximizes our return on investment



The Breakdown: * We translated the ratings_count < 50000 cap into "not yet peaked in mainstream fame."

We translated the ratings_count > 10000 floor into "definitively proven their elite quality."

We provided the "Why" (Return on Investment).

### 2. We have a marketing budget for this year. Which country should we prioritise and why?

We need a country that produces elite wine (Signal) with high consistency (Shield), but currently lacks the massive user adoption (Saturation) of the heavyweights.

In [9]:
query_path = Path("../queries/02_marketing_country.sql") 

top_country_marketing = execute_sql_file(query_path)

for country, total_clients, total_wines_country, average_total_ratings in top_country_marketing:
    print(f"{country=}, {total_clients=}, {total_wines_country=}, {average_total_ratings=}")

country='Espagne', total_clients=2264396, total_wines_country=94, average_total_ratings=4.44
country='Italie', total_clients=4270717, total_wines_country=383, average_total_ratings=4.43
country='France', total_clients=5973301, total_wines_country=331, average_total_ratings=4.45
country='États-Unis', total_clients=12273684, total_wines_country=74, average_total_ratings=4.49


Explanation to stakeholders

To maximize the return on our marketing budget, we analyzed our top-tier wine producing nations. While heavyweights like the United States, France, and Italy produce massive volumes of highly-rated wine, their markets are already heavily saturated with millions of existing users. Therefore, we recommend prioritizing Spain. Spain is a 'High-Quality Underdog.' It possesses the infrastructure to produce a massive variety of elite wines (94 distinct wines averaging a 4.44 rating), but currently has a fraction of the user base (2.2 million). Investing our budget here guarantees we are pushing a statistically proven, premium product into a market with massive untapped growth potential.


The Mechanics: This paragraph starts with the business goal (ROI), acknowledges the competition to build trust, presents the data (94 wines, 4.44 rating, 2.2m users), and closes with the strategic justification.

### 3. We would like to give awards to the best wineries. Come up with 3 relevant ones. Which wineries should we choose and why?

The awards:

1. The Time Capsule Award: Oldest Registered Bottle winery
2. The Prolific Elite Award: Highest volume production of wines
3. The Flawless Boutique: Masters of small-batch perfection

### The Time Capsule Award: Oldest Registered Bottle winery

In [10]:
query_path = Path("../queries/03_award_1.sql") 

award_1 = execute_sql_file(query_path)

print(award_1)

[('Tignanello', 'Quinta do Crasto Colheita Porto 2003', 2003)]


This award recognizes the winery that has preserved the deepest historical footprint in our platform. By analyzing our records and strictly filtering out incomplete data, we searched for the oldest authenticated bottle. The award goes to Tignanello. Their 2003 vintage stands as the oldest verified wine in our collection, representing a deep, enduring legacy in the wine market.

### The Prolific Elite Award: Highest volume production of wines

In [12]:
query_path = Path("../queries/03_award_2.sql") 

award_2 = execute_sql_file(query_path)

print(award_2)

[('Corte di Cama Sforzato di Valtellina', 1, 4.5)]


Due to constraints in the relational mapping between the wineries and wines tables, the maximum queryable production volume for any single entity in this dataset is 1 unit. When applying the 'Elite' criteria (average_ratings > 4.0), Corte di Cama Sforzato di Valtellina is selected as the top candidate that satisfies both the maximum available volume and the quality threshold.

### The Flawless Boutique: Masters of small-batch perfection

In [13]:
query_path = Path("../queries/03_award_3.sql") 

award_3 = execute_sql_file(query_path)

print(award_3)

[("Tenuta Tignanello 'Solaia'", 4.5, 1)]


Tenuta Tignanello 'Solaia' is awarded 'The Flawless Boutique' designation. By isolating the masters of small-batch perfection (wineries producing fewer than 5 wines), we found that Tenuta Tignanello 'Solaia' achieved the absolute highest average rating in the database. This validates their ability to consistently deliver top-tier quality across their limited production.

### 4. We have detected that a big cluster of customers like a specific combination of tastes. We have identified a few primary keywords that match this. We would like you to find all the wines that have those keywords:
- coffee
- toast
- green apple
- cream
- citrus
  
To ensure the accuracy of our selection, ensure that more than 10 users confirmed those keywords. Also, identify the group_name related to those keywords.

In [15]:
query_path = Path("../queries/04_flavor_groups.sql") 

flavor_groups = execute_sql_file(query_path)

for flavor in flavor_groups:
    print(flavor)

("'Giulio Ferrari' Riserva del Fondatore", 'citrus', 66, 'citrus_fruit')
("'Giulio Ferrari' Riserva del Fondatore", 'cream', 72, 'microbio')
("'Giulio Ferrari' Riserva del Fondatore", 'green apple', 15, 'tree_fruit')
("'Giulio Ferrari' Riserva del Fondatore", 'toast', 48, 'microbio')
("'Giulio Ferrari' Riserva del Fondatore", 'toast', 48, 'non_oak')
('25 anni Montefalco Sagrantino', 'coffee', 23, 'non_oak')
('25 anni Montefalco Sagrantino', 'coffee', 23, 'oak')
('5 Puttonyos Aszú (Blue Label)', 'citrus', 118, 'citrus_fruit')
('5 Puttonyos Aszú (Blue Label)', 'cream', 68, 'microbio')
('5 Puttonyos Aszú (Blue Label)', 'green apple', 12, 'tree_fruit')
('5 Puttonyos Aszú (Blue Label)', 'toast', 30, 'microbio')
('5 Puttonyos Aszú (Blue Label)', 'toast', 30, 'non_oak')
('5 Stelle Sfursat', 'coffee', 35, 'non_oak')
('5 Stelle Sfursat', 'coffee', 35, 'oak')
('50 & 50', 'coffee', 13, 'non_oak')
('50 & 50', 'coffee', 13, 'oak')
('60 Sessantanni Old Vines Primitivo di Manduria', 'coffee', 136, 'n

### 5. We would like to select wines that are easy to find all over the world. Find the top 3 most common grape all over the world and for each grape, give us the the 5 best rated wines.